src/Features/Core/Portable/CodeRefactoringHelpers.cs

```diff
@@ -3,12 +3,44 @@
 using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis
 {
     internal static class CodeRefactoringHelpers
     {
+
+        /// <summary>
+        /// <para>
+        /// Returns the relevant <see cref="SyntaxNode"/> instance if refactoring for <typeparamref name="TSyntaxNode"/> type
+        /// should be offered given the specified selection in document.
+        /// </para>
+        /// <para>
+        /// A <typeparamref name="TSyntaxNode"/> instance is returned if selection is zero-width and inside/touching a Token whose
+        /// direct parent is of type <typeparamref name="TSyntaxNode"/> or if a Node of said type is the smallest Node containing
+        /// the whole <paramref name="selection"/>. Otherwise returns <code>null</code>.
+        /// </para>
+        /// </summary>
+        public static async Task<TSyntaxNode> TryGetSelectedNode<TSyntaxNode>(
+            TextSpan selection, SyntaxNode root, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+
+            // TODO: Add handling selected whitespace before / after the same way as helper methods below -> refactor to public
+            // helper. 
+
+            var node = root.FindNode(selection) as TSyntaxNode;
+            if (node == null)
+            {
+                // e.g. "C LocalFunction[||](C c)" -> root.FindNode return ParameterList but we still want to return LocalFunctionNode
+                var identifier = await root.SyntaxTree.GetTouchingTokenAsync(selection.Start,
+                    token => token.Parent is TSyntaxNode, cancellationToken).ConfigureAwait(false);
+                node = identifier.Parent as TSyntaxNode;
```

Comment:
Thinking a bit more, we probably need to handle case where multiple nodes along a tree path have the same starting or ending token. For example, `[||]C LocalFunction(C c)`, identifier.Parent would give the `IdentifierNameSyntax` node for C, but it's parent local function node also has the same first token and so should be a candidate and all refactorings applicable to IdentifierNameSyntax and LocalFunctionStatementSyntax should be offered. Similarly, we have to handle the case where caret is just after the last token of multiple nodes.

Reply:
That gets slightly problematic, unfortunately because when do you walk the tree upwards & how far? Do you walk it only if you were on the edge of spans & as long as you stay on the edge (e.g. `[||]C LocalFunction(C c)`? Because you can't just walk upwards always, unless we want to show the refactoring in way more situations (e.g. (but not limited to) `"C[||] LocalFunction(C c)"`.

Going upwards as long as you're on the edge is (probably) seems pretty reasonable to me but it also doesn't seem entirely unreasonable to just not provide the refactoring in that case.

Reply:
FWIW: I think walking up left/right edge seems quite reasonable.

Reply:
> Going upwards as long as you're on the edge is (probably) seems pretty reasonable to me but it also doesn't seem entirely unreasonable to just not provide the refactoring in that case

Either way seems fine to me. We can start with your current implementation and iterate.

> BTW: There's similar issue with FindNode where it stops on innermost Node even if there's tie (it's possible to change it to outermost but to be truly correct I should add a version that goes through ties until it passes some predicate).

You should probably invoke it to get the innermost node for tie. When you make the walking up chain while on edge fix, it should fix this case. Alternatively, you can invoke FindNode twice (one with getInnermostNodeForTie = true and second with getInnermostNodeForTie = false) and then walk up the chain from innermost to outermost looking for matching node.

Reply:
As for touching, do we want to consider touching only at the beginning edge or also the end edge? I can see reasons for both but TBH climbing End edge doesn't seem all that intuitive.

E.g. should: 
```csharp
C LocalFunction(C c)
{
}[||]
``` 
also trigger?

Reply:
Yes, i would think it would trigger.  It seems totally intuitive to me that the user is 'selecting' the LocalFunction either being there or being here:

```c#
$$C LocalFunction(C c)
{
}
```

Note: i would consider this *after* doing the whitespace trimming.  So these would also count:

```c#
$$    C LocalFunction(C c)
    {
    }
```

Or

```c#
C LocalFunction(C c)
{
}    $$
```

I'm not sure if this should count though (but i'm not hugely opposed to it):

```c#
C LocalFunction(C c)
{
}

$$
```

Reply:
TBH I'm not so sure about trimming whitespace where there's no selection but just a carrot position. Considering 2, 3, and 4 as correct positions for refactoring opens issues (if we want to be consistent & general) in situations the current position could be associated with multiple nodes that might potentially be separated with (?a lot?) of whitespace. 

I'm all for being forgiving in case of selections, they're imprecise but for positions I'd say you need to be within/touching the node.

What I mean is: how often would a user expect refactoring with carrot here: 
```csharp
$$  C LocalFunction(C c)
    {
    }
```

but wouldn't be surprised if it wasn't when cursor is here:
```csharp
    C LocalFunction(C c)
$$  {
    }
```

(it might be just my experience where I use control+arrows for movement of my cursor so I'm always never within whitespace)

Reply:
Implemented without stripping whitespace for edges.